core/src/main/scala/caliban/parsing/parsers/SelectionParsers.scala

```diff
@@ -27,23 +27,18 @@ private[caliban] trait SelectionParsers extends ValueParsers {
   def namedType(implicit ev: P[Any]): P[NamedType] = P(name.filter(_ != "null")).map(NamedType(_, nonNull = false))
   def listType(implicit ev: P[Any]): P[ListType]   = P("[" ~ type_ ~ "]").map(t => ListType(t, nonNull = false))
 
-  def nonNullType(implicit ev: P[Any]): P[Type] = P((namedType | listType) ~ "!").map {
-    case t: NamedType => t.copy(nonNull = true)
-    case t: ListType  => t.copy(nonNull = true)
+  def type_(implicit ev: P[Any]): P[Type] = P((namedType | listType) ~ "!".!.?).map {
+    case (t: NamedType, nn) => if (nn.isDefined) t.copy(nonNull = true) else t
+    case (t: ListType, nn)  => if (nn.isDefined) t.copy(nonNull = true) else t
   }
-  def type_(implicit ev: P[Any]): P[Type]       = P(nonNullType | namedType | listType)
 
   def field(implicit ev: P[Any]): P[Field] =
-    P(Index ~ alias.? ~ name ~ arguments.? ~ directives.? ~ selectionSet.?).map {
-      case (index, alias, name, args, dirs, sels) =>
-        Field(
-          alias,
-          name,
-          args.getOrElse(Map()),
-          dirs.getOrElse(Nil),
-          sels.getOrElse(Nil),
-          index
-        )
+    P(Index ~ name ~ aliasOrName.? ~ arguments.? ~ directives.? ~ selectionSet.?).map {
+      case (index, alias, Some(name), args, dirs, sels) =>
```

Comment:
This is the main optimization. If the 3rd parser succeeded (indicated by `Some(_)`), then it means we have an alias. This is much faster because the 3rd parser (`aliasOrName`) contains the `:` assertion at the start of the string rather than the end